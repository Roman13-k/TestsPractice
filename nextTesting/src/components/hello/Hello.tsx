"use client";

import { useEffect, useState } from "react";

export default function Hello() {
  const [data, setData] = useState<null | object>(null);
  const [toggle, setToggle] = useState(false);
  const [value, setValue] = useState("");

  const handleToggle = () => {
    setToggle((prev) => !prev);
  };

  useEffect(() => {
    setTimeout(() => {
      setData({});
    }, 100);
  }, []);

  return (
    <div>
      <h1 data-testid='value-el'>{value}</h1>
      {data && <div style={{ color: "red" }}>data</div>}
      <h1>Hello, Next.js!</h1>
      {toggle && <div data-testid='toggle-div'>111</div>}
      <button data-testid='toggle-btn' onClick={handleToggle}>
        click me
      </button>
      <input onChange={(e) => setValue(e.target.value)} type='text' placeholder='input value ...' />
    </div>
  );
}
