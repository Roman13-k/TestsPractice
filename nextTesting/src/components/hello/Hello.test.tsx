import { fireEvent, render, screen } from "@testing-library/react";
import { userEvent } from "@testing-library/user-event";
import Hello from "./Hello";

describe("Hello component", () => {
  test("Render el", () => {
    render(<Hello />);
    expect(screen.getByText(/hello, next.js!/i)).toBeInTheDocument();
    expect(screen.getByRole("button")).toBeInTheDocument();
    expect(screen.getByPlaceholderText(/input value/i)).toBeInTheDocument();
  });

  test("async render", async () => {
    render(<Hello />);
    screen.debug();
    const asyncDiv = await screen.findByText(/data/i);
    expect(asyncDiv).toBeInTheDocument();
    expect(asyncDiv).toHaveStyle("color: rgb(255, 0, 0)");
    expect(asyncDiv).toMatchSnapshot();
    screen.debug();
  });

  test("BUTTON EVENT", () => {
    render(<Hello />);
    const toggleBtn = screen.getByTestId("toggle-btn");
    expect(screen.queryByTestId("toggle-div")).toBeNull();
    fireEvent.click(toggleBtn);
    expect(screen.queryByTestId("toggle-div")).toBeInTheDocument();
    fireEvent.click(toggleBtn);
    expect(screen.queryByTestId("toggle-div")).toBeNull();
  });

  test("INPUT EVENT", async () => {
    render(<Hello />);
    const input = screen.getByPlaceholderText(/input value/i);
    expect(screen.getByTestId("value-el")).toContainHTML("");
    // искусственное событие
    // fireEvent.change(input, {
    //   target: { value: "12345" },
    // });

    // близко к пользователю
    await userEvent.type(input, "12345");
    expect(screen.getByTestId("value-el")).toContainHTML("12345");
  });
});
