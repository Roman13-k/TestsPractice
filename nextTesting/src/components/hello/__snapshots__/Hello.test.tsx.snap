// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Hello component async render 1`] = `
<div
  style="color: red;"
>
  data
</div>
`;
